python/ray/data/read_api.py

```diff
@@ -2487,31 +2487,76 @@ def from_arrow_refs(
     )
 
 
+_DATABRICKS_SPARK_DATAFRAME_CHUNK_BYTES = 32 * 1024 * 1024
+
+
 @PublicAPI
 def from_spark(
     df: "pyspark.sql.DataFrame",
     *,
-    parallelism: Optional[int] = None,
+    parallelism: int = -1,
     override_num_blocks: Optional[int] = None,
 ) -> MaterializedDataset:
     """Create a :class:`~ray.data.Dataset` from a
     `Spark DataFrame <https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.html>`_.
 
     Args:
-        df: A `Spark DataFrame`_, which must be created by RayDP (Spark-on-Ray).
+        df: Any `Spark DataFrame` if you run Ray application with Ray on spark,
+            or a `Spark DataFrame`_ which must be created by RayDP (Spark-on-Ray).
         parallelism: This argument is deprecated. Use ``override_num_blocks`` argument.
         override_num_blocks: Override the number of output blocks from all read tasks.
             By default, the number of output blocks is dynamically decided based on
             input data size and available resources. You shouldn't manually set this
             value in most cases.
 
     Returns:
-        A :class:`~ray.data.MaterializedDataset` holding rows read from the DataFrame.
+        If you run Ray application with Ray on spark, it returns a ray dataset, the
+        data is cached in spark distributed cache system, but is not cached
+        in ray object store, otherwise it returns
+        a :class:`~ray.data.MaterializedDataset` holding rows read from the DataFrame.
     """  # noqa: E501
-    import raydp
+    from ray.util.spark.utils import is_in_databricks_runtime
 
     parallelism = _get_num_output_blocks(parallelism, override_num_blocks)
-    return raydp.spark.spark_dataframe_to_ray_dataset(df, parallelism)
+
+    if is_in_databricks_runtime():
+        from ray.data.datasource.spark_datasource import validate_requirements, SparkDatasource
+        from ray.util.spark.databricks_hook import get_databricks_function
+
+        validate_requirements()
+
+        # Ray on spark
+        datasource = SparkDatasource(df, _DATABRICKS_SPARK_DATAFRAME_CHUNK_BYTES)
+        if parallelism == -1:
+            parallelism = datasource.num_chunks
+        dataset = read_datasource(
+            datasource=datasource,
+            override_num_blocks=parallelism,
+        )
+
+        def cleanup_hook():
+            datasource.dispose_spark_cache()
+
+        dataset._cleanup_hook = cleanup_hook
+
+        try:
+            get_databricks_function("displayHTML")(
```

Comment:
Does this print out a logging? Where the logging is printed? Is it stdout?

Reply:
It is not logging, it is databricks-only function that shows a HTML widget in databricks notebook.

Reply:
Got it. Can you add a comment for this? This is not obvious for Ray users.